File: packages/llm-test-harness-python/llm_test_harness/metrics/word_count.py
from __future__ import annotations

from ..types import MetricFn, MetricScore


class WordCount(MetricFn):
    """Assert the response word count falls within a range.

    Example::

        Metrics.word_count(min=10)           # at least 10 words
        Metrics.word_count(max=100)          # enforce conciseness
        Metrics.word_count(min=20, max=200)  # bounded range
    """

    def __init__(self, *, min: int = 0, max: int = 0) -> None:
        self._min = min
        self._max = max if max > 0 else 2 ** 31

    @property
    def name(self) -> str:
        max_str = "∞" if self._max == 2 ** 31 else str(self._max)
        return f"WordCount({self._min}-{max_str})"

    def evaluate(self, text: str) -> MetricScore:
        count = len(text.split()) if text.strip() else 0
        passed = self._min <= count <= self._max
        reason: str | None = None
        if not passed:
            if count < self._min:
                reason = f"{count} words is below minimum {self._min}"
            else:
                reason = f"{count} words exceeds maximum {self._max}"
        return MetricScore(name=self.name, passed=passed, score=1.0 if passed else 0.0, reason=reason)

File: packages/llm-test-harness/src/metrics/word-count.ts
import type { MetricFn, MetricScore } from '../types.js'

/**
 * Asserts the response falls within a word count range.
 * Useful for checking verbosity (LLMs can be too brief or too verbose).
 *
 * @example
 * Metrics.wordCount({ min: 20 })         // at least 20 words
 * Metrics.wordCount({ max: 100 })        // no more than 100 words
 * Metrics.wordCount({ min: 10, max: 50}) // between 10 and 50 words
 */
export class WordCount implements MetricFn {
  readonly name: string
  private min: number
  private max: number

  constructor(options: { min?: number; max?: number }) {
    this.min = options.min ?? 0
    this.max = options.max ?? Infinity
    this.name = `WordCount(${this.min}-${this.max === Infinity ? '∞' : this.max})`
  }

  evaluate(text: string): MetricScore {
    const count = text.trim() === '' ? 0 : text.trim().split(/\s+/).length
    const pass = count >= this.min && count <= this.max

    let reason: string | undefined
    if (!pass) {
      if (count < this.min) reason = `${count} words is below minimum ${this.min}`
      else reason = `${count} words exceeds maximum ${this.max}`
    }

    return { name: this.name, pass, score: pass ? 1 : 0, reason }
  }
}
